wgpu-core/src/device/queue.rs

```diff
@@ -978,6 +979,62 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
                         baked
                             .initialize_texture_memory(&mut *trackers, &mut *texture_guard, device)
                             .map_err(|err| QueueSubmitError::DestroyedTexture(err.0))?;
+
+                        // Insert synthetic barriers to insert EXTERNAL barriers for any used external textures.
+                        {
+                            let mut used_external_textures = TextureUsageScope::new();
+
+                            let external_textures = baked
+                                .trackers
+                                .textures
+                                .used()
+                                .filter(|id| {
+                                    texture_guard[*id]
+                                        .inner
+                                        .as_raw()
+                                        .map(|inner| !unsafe { inner.is_external() })
+                                        .unwrap_or(false)
+                                })
+                                .collect::<Vec<_>>();
+
+                            for id in external_textures {
+                                // Get the last transition state for the texture.
+                                if let Some(last_transition) = baked
+                                    .trackers
+                                    .textures
+                                    .pending()
+                                    .filter(|pending| {
+                                        let (index, _, _) = id.0.unzip();
+                                        index == pending.id
+                                    })
+                                    .last()
+                                    .cloned()
```

Comment:
This feels like it can be made simpler without the need to manual iterate all these things.